collector/supervisord.go

```diff
@@ -0,0 +1,138 @@
+// Copyright 2015 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// +build !nosupervisord
+package collector
+
+import (
+	"flag"
+
+	"github.com/kolo/xmlrpc"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/log"
+)
+
+var (
+	supervisordURL = flag.String("collector.supervisord.url", "http://localhost:9001/RPC2", "XML RPC endpoint")
+)
+
+type supervisordCollector struct {
+	client         *xmlrpc.Client
+	upDesc         *prometheus.Desc
+	stateDesc      *prometheus.Desc
+	exitStatusDesc *prometheus.Desc
+	uptimeDesc     *prometheus.Desc
+}
+
+func init() {
+	Factories["supervisord"] = NewSupervisordCollector
+}
+
+func NewSupervisordCollector() (Collector, error) {
+
+	client, err := xmlrpc.NewClient(*supervisordURL, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	var (
+		subsystem  = "supervisord"
+		labelNames = []string{"name", "group"}
+	)
+	return &supervisordCollector{
+		client: client,
+		upDesc: prometheus.NewDesc(
+			prometheus.BuildFQName(Namespace, subsystem, "up"),
+			"Process Up",
+			labelNames,
+			nil,
+		),
+		stateDesc: prometheus.NewDesc(
+			prometheus.BuildFQName(Namespace, subsystem, "state"),
+			"Process State",
+			labelNames,
+			nil,
+		),
+		exitStatusDesc: prometheus.NewDesc(
+			prometheus.BuildFQName(Namespace, subsystem, "exit_status"),
+			"Process Exit Status",
+			labelNames,
+			nil,
+		),
+		uptimeDesc: prometheus.NewDesc(
+			prometheus.BuildFQName(Namespace, subsystem, "uptime"),
+			"Process Uptime",
+			labelNames,
+			nil,
+		),
+	}, nil
+}
+
+func (c *supervisordCollector) isRunning(state int) bool {
+	// http://supervisord.org/subprocess.html#process-states
+	const (
+		STOPPED  = 0
+		STARTING = 10
+		RUNNING  = 20
+		BACKOFF  = 30
+		STOPPING = 40
+		EXITED   = 100
+		FATAL    = 200
+		UNKNOWN  = 1000
+	)
+	switch state {
+	case STARTING:
+	case RUNNING:
+	case STOPPING:
```

Comment:
I don't think this is doing what you want. Assuming you want to return `true` in case the state is any of `STARTING`, `RUNNING` or `STOPPING`, then you have to to write it like this:

```go
switch state {
case STARTING, RUNNING, STOPPING:
  return true
}
return false
```

In your example, only `STOPPING` will become true, all other cases will fall through the switch statement and are covered by the closing `return false`.

Given there are no tests available for this collector, this might be a good first candidate :)